### DISCUSSION TITLE
How to print the function name instead of the address for ustack

### DISCUSSION THREAD

**Discussion by hsueh-yh** [bpftrace/bpftrace#3108] [General]
I tried:
```shell
bpftrace -e "profile:hz:99 {@stacks[comm,pid,ustack] = count();}"
```
and got:
```shell
...
@stacks[a.out, 172272,
    0x401210
    0x4014c1
    0x4015af
    0x401641
    0x40172b
    0x7f3393041493
    0x5541f689495641d7
]: 2
@stacks[a.out, 172272,
    0x4014d4
    0x4015af
    0x401641
    0x40172b
    0x7f3393041493
    0x5541f689495641d7
]: 2
@stacks[a.out, 172272,
    0x4014bc
    0x4015af
    0x401641
    0x401661
    0x401cb8
    0x401b19
    0x401eec
    0x401ec2
    0x401ea6
    0x7f3393a3fba3
]: 3
......
```

a.out is a simple c++ demo program compiled with
```shell
g++ -g -O0 test.cpp
```
bpftrace version: v0.13.0
os: centos8  linux 4.18

Expect:
ustack print the function name instead of address.

Question:
Did I missed any compilation parameters or packeges?

Thansk!

**Comment by ajor** [comment]
Symbolication is based off the symbol table of the target application. It doesn't look like you're doing anything wrong to me, but you could double check that symbols exist in your `a.out` in a couple of ways:

1. Look for "stripped" or "not stripped" in the output of `file a.out`.
2. Look at the symbols yourself with `nm -p a.out`. You'll either see a list of addresses and symbols or the message "no symbols".

If the symbols are fine then another thing to check is whether your process is still running when bpftrace is printing out the stack traces. bpftrace uses the PID to find the correct process to symbolicate against, so if the process no longer exists we won't find the symbol table.

**Reply by hsueh-yh** [reply]
Thanks for your swift reply!  I have checked steps 1 and 2 and can confirm:
1. The symbols have been properly compiled into ```a.out```.   
2. The process is a dead loop for testing, so it will always exist. 

Another key piece of infomation is that the program runs in a container, and the PID printed by bpftrace is diffrent from the PID printed by ```ps``` in the container. Will this cause bpftrace to fail to find the correct process or symbols by PID?

Thanks!

**Comment by ajor** [comment]
> Another key piece of infomation is that the program runs in a container, and the PID printed by bpftrace is diffrent from the PID printed by `ps` in the container. Will this cause bpftrace to fail to find the correct process or symbols by PID?

Is bpftrace running in the container too? That'll be the problem if so, as the bpftrace userland will see the namespaced PID while the BPF kernel program will get the PID as seen from the init namespace.

I'm actually working on a patch this week which will fix PID collection and symbolication for this case - will update you here when I submit it

**Reply by hsueh-yh** [reply]
@ajor  Has this issue been fixed?

**Reply by ajor** [reply]
@hsueh-yh I'm so sorry! I completely forgot about this. I'm literally working on the patch right now so hopefully I'll manage to get it finished this time before I get distracted again

**Reply by ajor** [reply]
I'll have to test it some more, but I've put up a draft fix now: #3428

**Comment by baiwfg2** [comment]
I have some findings. When I put `ustack` into map object just like you do, the symbol won't be printed. However, when I just use `printf("%s", ustack);` ，it will print symbol. NO IDEA why it's that.

**Reply by jordalgo** [reply]
@baiwfg2 Do you mind pasting your sample script and also the output of `bpftrace --info` ?

Ah, nm, I see your new discussion topic now: https://github.com/bpftrace/bpftrace/discussions/3614

**Comment by PoneyUHC** [comment]
I can reproduce also reproduce the same issue, where the symbol names are being stripped if the ustack is put in a map.
Let's consider printing the ustack for a program that does close syscalls called `closetest`
These scripts show the problem:
```
tracepoint:syscalls:sys_enter_close
/comm == "closetest"/
{
    @stack[pid] = ustack(perf, 5);
}
```
When exiting, the `@stack` map will hold ([unknown]) in place of the symbols.
However, printing the map once like in this script:
``` 
tracepoint:syscalls:sys_exit_close
/comm == "closetest"/
{
    @stack[pid] = ustack(perf, 5);
    print(@stack);
}
```
Will produce a complete (expected) output where all the symbols are resolved.

It looks like printing a ustack object once populated the structure under the hood, which is not done in the case where it's printed on exit when printing from maps.

**Reply by PoneyUHC** [reply]
For now, I have a dirty workaround that allows me to accumulate stacks in a map, and print a resolved version on exit.
The same map is not printed earlier (like shows above), because I don't want a confusion between the pre-print (to load symbols) and the final one.
```
BEGIN
{
    @id = 1;
    @init = 0;
}

tracepoint:syscalls:sys_exit_close
/comm == "closetest"/
{   
    if (@init == 0) {
        @trashstack = ustack(perf, 5);
        print(@trashstack);
        @init = 1;
    }
    
    @stacks[@id] = ustack(perf, 5);
    @id++;
}
```
It looks like any ustack print, regardless of the map, will load the symbols for exit map printing to have it.

**Reply by jordalgo** [reply]
@PoneyUHC Thanks for reporting. I have to dig into this a little more but here is my guess about what is happening. The process which bpftrace is getting the ustack for is exiting before the script exists so by the time we need to print those symbols we can't lookup the process to get them. Your workaround is effective because we probably cache those symbols when you print them right away with `print(@trashstack)` so that when the script does exist we still have access to them. A useful feature is probably sending any `ustack` into userspace for symbolication caching (regardless if it's being printed at the time). But again, this is just a guess. I'll look into it.

**Reply by jordalgo** [reply]
@PoneyUHC I forgot, if you're in control of this program, you can have bpftrace launch it so it doesn't die before you collect symbols via the `-c` flag.

**Reply by PoneyUHC** [reply]
> @PoneyUHC I forgot, if you're in control of this program, you can have bpftrace launch it so it doesn't die before you collect symbols via the `-c` flag.

Thanks for the tip ! It will definitely help for my workflow (PhD work), where I do control the programs :)

**Reply by jordalgo** [reply]
@PoneyUHC You uncovered an issue that has been around for a while and should be fixed here: https://github.com/bpftrace/bpftrace/pull/5215
